### PR TITLE
FEAT: Implement `??`

### LINKED ISSUE #938: Support ES2020 Nullish Coalescing Operator
Proposal: https://github.com/tc39/proposal-nullish-coalescing
MDN: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator

### PR DESCRIPTION
**Pull request by 0xe**
Closes #938

### CODE REVIEW

File: rhino/src/main/java/org/mozilla/javascript/Token.java [hunk 1]
```diff
@@ -227,6 +227,7 @@ public static enum CommentType {
             TEMPLATE_LITERAL_SUBST = 173, // template literal - substitution
             TAGGED_TEMPLATE_LITERAL = 174, // template literal - tagged/handler
             DOTDOTDOT = 175, // spread/rest ...
+            NULLISH_COALESCING = 176, // nullish coalescing (??)
             LAST_TOKEN = 175;
```
Comment (andreabergia) [inline rhino/src/main/java/org/mozilla/javascript/Token.java:231]: This needs to be updated too

Comment (0xe) [reply rhino/src/main/java/org/mozilla/javascript/Token.java:231]: Will do, missed that!

File: tests/src/test/java/org/mozilla/javascript/tests/NullishCoalescingOpTest.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import org.junit.Assert;
+import org.junit.Test;
+import org.mozilla.javascript.Context;
+import org.mozilla.javascript.Scriptable;
+
+public class NullishCoalescingOpTest {
+
+    @Test
+    public void testNullishColascingBasic() {
+        try (Context cx = Context.enter()) {
```
Comment (rbri) [inline tests/src/test/java/org/mozilla/javascript/tests/NullishCoalescingOpTest.java:12]: please check if you can use org.mozilla.javascript.tests.Utils#assertWithAllOptimizationLevelsES6 or similar for your test cases. At least make sure your tests are processed with all optimization levels.

**Comment by rbri** [comment]
@0xe maybe https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_OR_assignment and https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_AND_assignment are doable in a similar way.... ;-)

**Comment by p-bakker** [comment]
You'll want to uncomment the `language/expressions/coalesce` section in test262.properties, so all the tests related to this feature will run

**Comment by p-bakker** [comment]
> @0xe maybe https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_OR_assignment and https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_AND_assignment are doable in a similar way.... ;-)

While at it, why not toss in the optional chaining operator 🤪 See #936 and #937

**Comment by nabacg** [comment]
still WIP #1593 but I couldn't help myself to give it a try!

**Comment by p-bakker** [comment]
Pipeline fails cause not all tests from `language/expressions/coalesce` pass

**Comment by p-bakker** [comment]
@rbri I assume you meant ||= and &&= instead of |= and &=, as I think the latter two bitwise operators (which you linked to) are already supported

**Comment by andreabergia** [comment]
> @0xe maybe https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_OR_assignment and https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_AND_assignment are doable in a similar way.... ;-)

https://github.com/mozilla/rhino/pull/1597 :)

**Comment by 0xe** [comment]
> Pipeline fails cause not all tests from `language/expressions/coalesce` pass

Thanks, I've fixed the failing test.

**Comment by gbrail** [comment]
This looks good to me -- any thoughts from anyone else?

One minor question -- the addition of a "NULLISH_COALESCING" token is a name we'll have to live with for a long time. Are we sure that we like that name -- does it match how people talk about this feature elsewhere in the JavaScript world? Or is another name more appropriate.

+1 for adding a French translation, BTW!

**Comment by gbrail** [comment]
Also, if others are OK then please rebase it so that we can merge it. Thanks!

**Comment by 0xe** [comment]
> This looks good to me -- any thoughts from anyone else?
> 
> One minor question -- the addition of a "NULLISH_COALESCING" token is a name we'll have to live with for a long time. Are we sure that we like that name -- does it match how people talk about this feature elsewhere in the JavaScript world? Or is another name more appropriate.
> 
> +1 for adding a French translation, BTW!

Thanks. We picked `NULLISH_COALESCING` based on what [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing) calls it. Happy to change if another name is more commonly used in the community.

**Comment by 0xe** [comment]
> Also, if others are OK then please rebase it so that we can merge it. Thanks!

I've rebased it against master.

**Pull request merged**